### PR TITLE
Prototype AOT generation as an MSBuild task

### PR DESCRIPTION
**Pull request by Perksey**
### Purpose of this PR

This PR aims to refactor AOT into a stable and production-ready component, by integrating the AOT generator into the MSBuild pipeline and including little improvements (such as version watermarks) to ensure its ongoing stability.

### Testing status

TBA.

### CODE REVIEW

**Comment by Perksey** [comment]
Ready for review, subsequent PRs to make the task target .NET Standard instead of .NET Framework will follow - it's a large undertaking and we need a base for now.

File: AdvancedDLSupport.AOT.Tasks/AdvancedDLSupport.AOT.Tasks.targets [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    <AOTOutputDirectory>adl-aot</AOTOutputDirectory>
+  </PropertyGroup>
+  <Target Name="IncludeAOTAssemblies" BeforeTargets="BeforeBuild">
+    <ItemGroup>
+      <AdditionalFiles Include="$(AOTOutputDirectory)/*.dll" InProject="false" CopyToOutputDirectory="PreserveNewest" />
+    </ItemGroup>
+  </Target>
+  <Target Name="BuildAOTAssemblies" DependsOnTargets="Build">
+    <GenerateDLDynamicAssemblies Input="$(TargetPath)" Output="$(ProjectDir)/$(AOTOutputDirectory)" />
+    <Copy SourceFiles="$(ProjectDir)/$(AOTOutputDirectory)/*.dll"
```
Comment (Nihlus) [inline AdvancedDLSupport.AOT.Tasks/AdvancedDLSupport.AOT.Tasks.targets]: Would this leave files in the ProjectDir?

File: AdvancedDLSupport.AOT.Tasks/GenerateDLDynamicAssembliesTask.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    /// <summary>
+    /// Represents an MSBuild task that can pre-generate implementations of <see cref="NativeLibraryBase"/>s.
+    /// </summary>
+    [UsedImplicitly]
+    public class GenerateDLDynamicAssembliesTask : Task
+    {
+        /// <summary>
+        /// Gets or sets the input file to pass to the task.
+        /// </summary>
+        [Required]
```
Comment (Nihlus) [inline AdvancedDLSupport.AOT.Tasks/GenerateDLDynamicAssembliesTask.cs]: Annotate nullablilty in this class.

File: AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj [hunk 3]
```diff
@@ -28,12 +27,12 @@
...
-    <PackageReference Include="CommandLineParser" Version="2.2.1" />
-    <PackageReference Include="NLog" Version="4.5.3" />
-    <PackageReference Include="StrictEmit" Version="2.4.2" PrivateAssets="Compile;Build;Analyzers;ContentFiles"/>
+    <PackageReference Include="CommandLineParser" Version="2.5.0" />
+    <PackageReference Include="NLog" Version="4.6.5" />
+    <PackageReference Include="StrictEmit" Version="2.4.2" PrivateAssets="Compile;Build;Analyzers;ContentFiles" />
     <PackageReference Include="JetBrains.Annotations" Version="11.1.0" PrivateAssets="all" />
     <PackageReference Include="JetBrainsAnnotations.Fody" Version="2.2.0" PrivateAssets="all" />
-    <PackageReference Include="System.Reflection.Emit" Version="4.3.0" />
+    <PackageReference Include="System.Reflection.Emit" Version="4.6.0-preview6.19303.8" />
```
Comment (Nihlus) [inline AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj]: Is using a preview version a requirement? Would prefer not to.

File: AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs [hunk 4]
```diff
@@ -176,11 +169,11 @@ public string Build([NotNull] string outputPath)
             var automaticInterfaces = new List<Type>();
             foreach (var sourceAssembly in SourceAssemblies)
             {
-                _log.Info($"Scanning {sourceAssembly.GetName().Name}...");
                 foreach (var automaticInterface in sourceAssembly.ExportedTypes.Where(t => t.HasCustomAttribute<AOTTypeAttribute>()))
                 {
                     automaticInterfaces.Add(automaticInterface);
-                    _log.Info($"Discovered {automaticInterface.Name}.");
+
+                    // TODO logging
```
Comment (Nihlus) [inline AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: Why was the logging removed?

File: AdvancedDLSupport.AOT.Tasks/AdvancedDLSupport.AOT.Tasks.targets (hunk 1 above)
Comment (Perksey) [reply AdvancedDLSupport.AOT.Tasks/AdvancedDLSupport.AOT.Tasks.targets]: Yes. Will change.

Comment (Perksey) [reply AdvancedDLSupport.AOT.Tasks/AdvancedDLSupport.AOT.Tasks.targets]: Fixed.

File: AdvancedDLSupport.AOT.Tasks/GenerateDLDynamicAssembliesTask.cs (hunk 2 above)
Comment (Perksey) [reply AdvancedDLSupport.AOT.Tasks/GenerateDLDynamicAssembliesTask.cs]: Fixed.

File: AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs (hunk 4 above)
Comment (Perksey) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: MSBuild can't use NLog. Mind, I could have it so NLog can be used only if available. Will fix now.

Comment (Nihlus) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: What's the effect if used through an MSBuild task? A crash, nothing happening...?

Comment (Perksey) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: The user will just never see the logs I think.

Comment (Nihlus) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: Might as well not remove it, then, since it'd work when running from the command line.

Comment (Perksey) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: Fixed.

File: AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj (hunk 3 above)
Comment (Perksey) [reply AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj]: Ah this was me trying to figure out if corefx supported RunAndSave at all, will fix.

File: AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs (hunk 4 above)
Comment (Nihlus) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: That being said, is it possible to inject the logging instance used in the task?

Comment (Perksey) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: Yes, but it'd require us referencing MSBuild from the console application.

File: AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj (hunk 3 above)
Comment (Perksey) [reply AdvancedDLSupport.AOT/AdvancedDLSupport.AOT.csproj]: Fixed.

File: AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs (hunk 4 above)
Comment (Nihlus) [reply AdvancedDLSupport.AOT/PregeneratedAssemblyBuilder.cs]: Let's leave it as is for now, then. It's not critical.

**Comment by Perksey** [comment]
Rectified review comments :)

**Comment by Nihlus** [comment]
Can you rebase this on master and ensure the tests pass before I review it?

**Comment by Perksey** [comment]
This PR will need to be closed for a short while as we pursue a large refactor (moving to Cecil)

**Pull request closed without merging**